airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py

```diff
@@ -0,0 +1,252 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+import os
+import urllib
+from pathlib import Path
+from typing import Any, Mapping, Optional, Tuple, Union
+
+import requests
+import requests_cache
+from airbyte_cdk.sources.http_config import MAX_CONNECTION_POOL_SIZE
+from airbyte_cdk.sources.streams.call_rate import APIBudget, CachedLimiterSession, LimiterSession
+from airbyte_cdk.utils.constants import ENV_REQUEST_CACHE_PATH
+from airbyte_cdk.utils.traced_exception import AirbyteTracedException
+from requests.auth import AuthBase
+
+from .error_handlers import (
+    BackoffStrategy,
+    DefaultBackoffStrategy,
+    ErrorHandler,
+    ErrorMessageParser,
+    ErrorResolution,
+    HttpStatusErrorHandler,
+    JsonErrorMessageParser,
+    ResponseAction,
+)
+from .exceptions import DefaultBackoffException, RequestBodyException, UserDefinedBackoffException
+from .rate_limiting import http_client_default_backoff_handler, user_defined_backoff_handler
+
+BODY_REQUEST_METHODS = ("GET", "POST", "PUT", "PATCH")
+
+
+class HttpClient:
+
+    _DEFAULT_MAX_RETRY = 5
+    _DEFAULT_RETRY_FACTOR = 5
+    _DEFAULT_MAX_TIME = 60 * 10
+
+    def __init__(
+        self,
+        name: str,
+        logger: logging.Logger,
+        error_handler: Optional[ErrorHandler] = None,
+        api_budget: Optional[APIBudget] = None,
+        session: Optional[Union[requests.Session, requests_cache.CachedSession]] = None,
+        authenticator: Optional[AuthBase] = None,
+        use_cache: bool = False,
+        backoff_strategy: BackoffStrategy = DefaultBackoffStrategy(),
```

Comment:
I think we want to change this to:
```
backoff_strategy: Optional[BackoffStrategy] = None
...
self._backoff_strategy = backoff_strategy or DefaultBackoffStrategy()
```

One issue w/ instantiating instances within constructor params is that it can end up getting shared between individual instantiations of the `HttpClient`. I'm not 100% sure if we'll run into this problem, but I can bet that it would be an incredibly subtle bug for us to miss if it did. It's also more consistent to how we instantiate APIBudget

Same. for the `ErrorMessageParser` as well.

For reference: https://stackoverflow.com/questions/4841782/python-constructor-and-default-value

Reply:
Interesting. Thanks for the note and the reference. Updated.